lambda-buffers-codegen/src/LambdaBuffers/Common/Types.hs

```diff
@@ -0,0 +1,158 @@
+                 necessary at this time. -}
+  | RefP Pat {- still unclear on how to handle these here -}
+  | AppP Pat Pat {- Pattern for Type applications -}
+  | {- This last one is a bit special. This represents a complete type declaration.
+
+       The first Pat should be instantiated to `Name l` where l is a concrete name.
+
+       The second Pat should be instantiated to a Pat-List (using :*/Nil) which only contains Names.
+
+       The final Pat should be instantiated to a Pat body.
+
+       In some languages, parts of this may be ignored. E.g. in Rust the type name doesn't matter (we use the constr name of the
+       outermost inner sum for constructing types). -}
+    DecP Pat Pat Pat
+  deriving (Show, Eq, Ord)
+
+infixr 5 :*
+
+-- Pattern synonyms. These VASTLY improve readability
+pattern Int :: Pat
+pattern Int = RefP (Name "Int")
+
+pattern String :: Pat
+pattern String = RefP (Name "String")
+
+pattern Bool :: Pat
+pattern Bool = RefP (Name "Bool")
+
+pattern List :: Pat -> Pat
+pattern List t = AppP (RefP (Name "List")) t
+
+pattern Maybe :: Pat -> Pat
+pattern Maybe t = AppP (RefP (Name "Maybe"))  t
+
+pattern Map :: Pat -> Pat -> Pat
+pattern Map k v = AppP (AppP (RefP (Name "Map")) k) v
+
+pattern Either :: Pat -> Pat -> Pat
+pattern Either l r = AppP (AppP (RefP (Name "Either")) l) r
+
+pattern RecFields :: Pat -> Pat -> Pat -> Pat
+pattern RecFields l x xs = RecP ((l := x) :* xs)
+
+pattern ProdArgs :: Pat -> Pat -> Pat
+pattern ProdArgs x xs = ProdP (x :* xs)
+
+pattern RecBody :: Pat -> Pat -> Pat
+pattern RecBody constr fields = SumP (constr := RecP fields :* Nil)
+
+pattern ProdBody :: Pat -> Pat -> Pat
+pattern ProdBody constr fields = SumP ((constr := ProdP fields) :* Nil)
+
+pattern Rec :: Pat -> Pat -> Pat -> Pat -> Pat
+pattern Rec tyName tyVars constr fields =
+  DecP tyName tyVars (SumP (constr := RecP fields :* Nil))
+
+pattern Product :: Pat -> Pat -> Pat -> Pat -> Pat
+pattern Product tyName tyVars constr args =
+  DecP tyName tyVars (SumP (constr := ProdP args :* Nil))
+
+pattern SumBody :: Pat -> Pat -> Pat -> Pat
+pattern SumBody l x xs = SumP ((l := x) :* xs)
+
+pattern Sum :: Pat -> Pat -> Pat -> Pat
+pattern Sum tyName tyVars pat =
+  DecP tyName tyVars pat
+
+pattern Void :: Pat
+pattern Void = SumP Nil
+
+-- we probably don't need these _ functions
+_list :: Pat
+_list = RefP (Name "List")
+
+_maybe :: Pat
+_maybe = RefP (Name "Maybe")
+
+_map :: Pat
+_map = RefP (Name "Map")
+
+_either :: Pat
+_either = RefP (Name "Either")
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+toProd :: [Pat] -> Pat
+toProd = ProdP . foldr (:*) Nil
+
+toRec :: [Pat] -> Pat
+toRec = RecP . foldr (:*) Nil
+
+toSum :: [Pat] -> Pat
+toSum = SumP . foldr (:*) Nil
+
+{- Utility functions. Turn a list of types into a product/record/sum type.
+-}
+
+
+-- misc utilities, break out into a Utils module if this grows too much
+
+-- whyyyyyy isn't this in prelude
+for :: [a] -> (a -> b) -> [b]
```

Comment:
It is
 
```
[for :: (Traversable t, Applicative f) => t a -> (a -> f b) -> f (t b)](https://hackage.haskell.org/package/base/docs/Data-Traversable.html#v:for) 
```
Use `Identity` as the Applicative ?

Reply:
Oh, neat, thanks!